### PR TITLE
Change management of test cfg to better support json projects

### LINKED ISSUE #4508: JsonProject projects not resolving sysroot symbols behind #[cfg(not(test))]
I'm seeing a difference between a standard `Cargo.toml` project and a `JsonProject` in how the sysroot packages are being processed, that I haven't been able to resolve.

In a project with `use std::fmt::Debug`, the `Debug` symbol doesn't get resolved (is flagged as `unresolvedReference` by rust-analyzer.  But `std::fmt` is.

Eventually I found this in `libcore/lib.rs`:
Screen Shot 2020-05-15 at 4 12 13 PM

Where it appears that the module isn't being found/processed correctly, and the same with any of the other modules preceded by `#[cfg(not(test))]`

In a Cargo.toml project, it is:
Screen Shot 2020-05-15 at 4 12 24 PM

I note that here:
https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_project_model/src/lib.rs#L323

The `test` atom is specifically being removed from sysroot crates for Cargo projects, but that isn't happening for JsonProject workspaces.

This comment implies that `test` is being stripped for the default CfgOptions (based on the `rustc` invocation, but doesn't perform that operation in this fn:
https://github.com/rust-analyzer/rust-analyzer/blob/38e8f35855efac144373c1b5aab3af050e47e594/crates/ra_project_model/src/lib.rs#L550

In general, the lines I referenced above, in ra_project_model/src/lib.rs imply that JsonProject should be providing all the features/atoms for the toolchain that it's being used with, especially in the case of cross-compilation, where the `rust` invoked at `get_rustc_cfg_options()` may not be the right one for the project.

### PR DESCRIPTION
**Pull request by Nashenas88**
This helps support json projects where they can decide whether to add the `test` cfg or not. One alternative is to add support for marking json project crates as a sysroot crate, and adding logic to remove the `test` cfg in those cases. In my opinion, that option gives less flexibility to json projects and leads to more functionality that needs to be maintained.

Fixes #4508 
cc @woody77

### CODE REVIEW

**Comment by woody77** [comment]
RE sysroot crates, it seems like they're handled just differently enough that making that explicit would be a good thing.  OTOH, with something like this, the rust-analyzer.json mechanism lets the file generator have a great deal of control over how RA interprets each crate.  And also gives the ability to better support cross-compiler situations (because the json file can provide all of the various `target_foo=bar` values (`target_arch`, `target_os`, `target_family`, etc).

One thing we (@Nashenas88 and I) should look into is what happens when the same crate is included multiple times with different cfgs...  I'm guessing there's a first-one/last-one-wins sort of situation, but I haven't looked into how an individual file is linked to the crate in the crate graph...

**Comment by matklad** [comment]
Hm, yeah, agree that the current handling of cfgs doesn't make sense. For JSON projects, we should just take the provided CFGs as is. 

I think these `default_cfg_options` are a premature generalization. Let's remove this argument from the `to_crate_graph` and instead just hard-code the current logic into the `ProjectWorkspace::Cargo` branch. 

>should look into is what happens when the same crate is included multiple times with different cfgs... 

This should in theory be fine, we do not assume that each file belongs to just one crate and this is how we want to handle `#[cfg(test)]` and such long term (by creating two crates in the graph, one with and one without `#[cfg(test)]`). We don't actively do this in practice yet on a significant scale, so there might be some issue. 

One specific one is that, when a user asks, eg, a completion in a certain file, we don't try to be smart about figuring out the crate in which context this should happen. Rather, we just pick the first crate which includes this file.

**Comment by Nashenas88** [comment]
@matklad I implemented your request.

File: crates/ra_project_model/src/lib.rs [hunk 1]
```diff
@@ -250,7 +250,7 @@ impl ProjectWorkspace {
 
     pub fn to_crate_graph(
         &self,
-        default_cfg_options: &CfgOptions,
+        target: Option<&String>,
```
Comment (matklad) [inline crates/ra_project_model/src/lib.rs:253]:
```suggestion
        target: Option<&str>,
```

`&String`, `&Vec`, `&Box` etc are sort-of useless double indirection types. 

r=me with this change

bors d+

**Pull request merged**